api-resources.md

```diff
@@ -0,0 +1,414 @@
+enum {
+  continued(0),
+  finished(1),
+  failed(2),
+  (255),
+} PrepareStepState;
+
+struct {
+  ReportID report_id;
+  PrepareStepState prepare_step_state;
+  select (PrepareStep.prepare_step_state) {
+    case continued: opaque prep_msg<0..2^32-1>; /* VDAF preparation message */
+    case finished: Empty;
+    case failed: ReportShareError;
+  }
+} PrepareStep;
+
+struct {
+  u16 round;
+  PrepareStep prepare_steps<1..2^32-1>;
+} AggregateJob
+```
+
+See "POST", below, for explanation and justification of the `round` field.
+
+#### Required HTTP methods
+
+##### GET/HEAD
+
+The response body is a `struct AggregateJob` representing the current state of the reports in the job.
+
+##### PUT
+
+PUT to an aggregate job is how the leader creates an aggregate job in the helper. The body is an `AggregateJobPutReq`:
+
+```
+struct {
+  ReportMetadata metadata;
+  opaque public_share<0..2^32-1>;
+  HpkeCiphertext encrypted_input_share;
+} ReportShare;
+
+struct {
+  QueryType query_type;
+  select (PartialBatchSelector.query_type) {
+    case time_interval: Empty;
+    case fixed_size: BatchID batch_id;
+  };
+} PartialBatchSelector;
+
+struct {
+  opaque agg_param<0..2^16-1>;
+  PartialBatchSelector part_batch_selector;
+  ReportShare report_shares<1..2^32-1>;
+} AggregateJobPutReq;
+```
+
+If successful, the helper's response is a `struct AggregateJob` where `prepare_steps` contains the helper's first-round VDAF prepare messages for the report shares in `AggregateJobPutReq`, or errors if `vdaf.prepare` failed for any report.
+
+This is analogous to DAP-02's `POST /aggregate` with an `AggregateInitReq`.
+
+This request is idempotent, so if a helper receives multiple PUT requests for some aggregation job and the members of `AggregateJobPutReq` don't change, the helper can return 200 OK.
+
+The idea here is that all the values in `struct AggregateJobPutReq` never need to be transmitted again during the handling of an aggregate job, which is why the message is not `struct AggregateJob`.
+
+##### POST
+
+POST to an aggregate job is how the leader steps an aggregate job. The request body is a `struct AggregateJob` where `prepare_steps` contains the leader's current-round prepare messages, and `round` is the number of the current round. The response is a `struct AggregateJob` where `prepare_steps` contains the helper's next-round prepare messages. The response's `round` field will be one more than the request's `round`.
+
+We use a POST here because this request is _not_ idempotent. Suppose the leader sends `POST /tasks/task-id/aggregate_jobs/aggregation-job-id` where the body has `round = 1`, and that this is successfully handled by the helper. Besides generating the response, this will have the side effect of advancing the state in the helper to round 2. If the leader were to resend the `POST /tasks/task-id/aggregate_jobs/aggregation-job-id` with `round = 1`, then the request should be rejected by the helper, because it has advanced to the next round.
```

Comment:
I like how this proposal eliminates an extra `GET` roundtrip. My intuition though is that the helper has to hold on not just to a hash of the previous request, but possibly also its response from the previous round. I've got to try to sketch out some failure scenarios in many-round VDAFs to see how this plays out.

Reply:
I think I get it now -- your scheme assumes that the leader would record its state before trying to advance state in the helper, which means that in the case of a crash, the helper can never be ahead of the leader.

I think that same assumption eliminates the need for the `GET` in the `round` solution. Suppose the helper is on round _n_. After recovering from a crash, the leader sends `POST` to the aggregate job with either:

- `round = n`: helper has seen this message before the leader crashed and sends its current state (round _n+1_ prepare messages). Protocol continues from there. 
- `round = n+1`: helper advances its state to the next round. Protocol continues from there.
- any other value of `round`: the leader and helper are too far out of sync. The aggregate job must be abandoned. The leader could recover by creating a new aggregate job with the same parameters, essentially going back to the first round. 

So I think either notion (request hash or round number) can achieve the functionality we want. Question is: which is "better"?

Reply:
> Where would the helper hashing the leader's PUT request help?

The only purpose of storing the (hash of) the previous request is to ensure that the response the Helper gives is coherent (i.e. matches the request).

While you are correct that the Leader should (if properly following DAP) always send the same request, there is nothing practically ensuring this if the Helper doesn't check somehow. A sufficiently-buggy Leader could send any request at all, and without the hash (or some other mechanism) to allow the request to be verified, the Helper would give the answer a non-buggy Leader would have received rather than the error the actually-buggy Leader should receive.

If folks don't think it's valuable to avoid this scenario, storage of the (hash of) the request can be dropped. IMO, I wouldn't recommend this (it's unlikely to come up in practice, but if it does, the results will likely be _very_ confusing to operators).

Reply:
Yeah, I think you're on to something. I had hoped that cases where the leader is sending the wrong prepare messages after a restart would be caught by failures in the helper's `Vdaf.prep_next` call, but at best they'd be caught on the next round.

What do you think about considering this fault recovery case separately from the resource oriented API change? The PR that introduces that into the protocol text will already be quite large, and this scenario that we are discussing is nuanced enough that it would benefit from having a spotlight shone on it, which is to say, a PR dedicated to making just that change.

Reply:
Sure -- I think the issues are separable. Probably the resource-oriented API should be specified first, then can be tweaked if necessary to achieve idempotency or otherwise solve the fault-recovery issue.

Reply:
It occurs to me it might wiser to leave it up to the implementation: the fault-recovery logic `MUST` check the request's round number, and `SHOULD`(?) check the entire request [worded to allow checking equality or by hash or some other technically-approximate method]. WDYT?